src/Symfony/Component/AssetMapper/ImportMap/RemotePackageDownloader.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\AssetMapper\ImportMap;
+
+use Symfony\Component\AssetMapper\ImportMap\Resolver\PackageResolverInterface;
+
+/**
+ * @final
+ */
+class RemotePackageDownloader
+{
+    private array $installed;
+
+    public function __construct(
+        private readonly ImportMapConfigReader $importMapConfigReader,
+        private readonly PackageResolverInterface $packageResolver,
+        private readonly string $vendorDir,
+    ) {
+    }
+
+    /**
+     * Downloads all packages.
+     *
+     * @return string[] The downloaded packages
+     */
+    public function downloadPackages(callable $progressCallback = null): array
+    {
+        try {
+            $installed = $this->loadInstalled();
+        } catch (\InvalidArgumentException) {
+            $installed = [];
+        }
+        $entries = $this->importMapConfigReader->getEntries();
+        $remoteEntriesToDownload = [];
+        $newInstalled = [];
+        foreach ($entries as $entry) {
+            if (!$entry->isRemotePackage()) {
+                continue;
+            }
+
+            // if the file exists at the correct version, skip it
+            if (
+                isset($installed[$entry->importName])
+                && $installed[$entry->importName]['version'] === $entry->version
+                && file_exists($this->vendorDir.'/'.$installed[$entry->importName]['path'])
+            ) {
+                $newInstalled[$entry->importName] = $installed[$entry->importName];
+                continue;
+            }
+
+            $remoteEntriesToDownload[$entry->importName] = $entry;
+        }
+
+        if (!$remoteEntriesToDownload) {
+            return [];
+        }
+
+        $contents = $this->packageResolver->downloadPackages($remoteEntriesToDownload, $progressCallback);
+        $downloadedPackages = [];
+        foreach ($contents as $package => $content) {
+            if (!isset($remoteEntriesToDownload[$package])) {
```

Comment:
I'd do the opposite. Iterate over $remoteEntriesToDownload to ensure every $entry has been downloaded. 

Or unset $remoteEntriesToDownload each iteration and throw after the loop if it still contains entries ?